Bio/Align/exonerate.py

```diff
@@ -0,0 +1,650 @@
+                        assert dummy == 0
+                    if query_step == 0:
+                        assert target_step > 0
+                        i += 1
+                        dummy, query_step = steps[i]
+                        assert dummy == 0
+                    operation = operations[i]
+                    assert operation == "U"
+                    operation = "N"  # Non-equivalenced region; exonerate definition
+                elif operation == "S":  # Split codon
+                    step = target_step
+                elif operation == "F":  # Frame shift
+                    step = target_step
+                else:
+                    raise ValueError("Unknown operation %s" % operation)
+                words.append(operation)
+                words.append(str(query_step))
+                words.append(str(target_step))
+                i += 1
+        line = " ".join(words) + "\n"
+        return line
+
+
+class AlignmentIterator(interfaces.AlignmentIterator):
+    """Alignment iterator for the Exonerate text, cigar, and vulgar formats.
+
+    Each line in the file contains one pairwise alignment, which are loaded
+    and returned incrementally.  Alignment score information such as the number
+    of matches and mismatches are stored as attributes of each alignment.
+    """
+
+    def __init__(self, source):
+        """Create an AlignmentIterator object.
+
+        Arguments:
+         - source   - input data or file name
+
+        """
+        super().__init__(source, mode="t", fmt="PSL")
+        stream = self.stream
+        self.program = "exonerate"
+        line = next(stream)
+        prefix = "Command line: "
+        assert line.startswith(prefix)
+        commandline = line[len(prefix) :].strip()
+        assert commandline.startswith("[")
+        assert commandline.endswith("]")
+        self.commandline = commandline[1:-1]
+        line = next(stream)
+        prefix = "Hostname: "
+        assert line.startswith(prefix)
+        hostname = line[len(prefix) :].strip()
+        assert hostname.startswith("[")
+        assert hostname.endswith("]")
+        self.hostname = hostname[1:-1]
+
+    @staticmethod
+    def _parse_cigar(words):
+        query_id = words[0]
+        query_start = int(words[1])
+        query_end = int(words[2])
+        query_strand = words[3]
+        target_id = words[4]
+        target_start = int(words[5])
+        target_end = int(words[6])
+        target_strand = words[7]
+        score = int(words[8])
+        target_seq = Seq(None, length=target_end)
+        query_seq = Seq(None, length=query_end)
+        target = SeqRecord(target_seq, id=target_id)
+        query = SeqRecord(query_seq, id=query_id)
+        qs = 0
+        ts = 0
+        n = (len(words) - 8) // 2
+        coordinates = numpy.empty((2, n + 1), int)
+        coordinates[0, 0] = ts
+        coordinates[1, 0] = qs
+        for i, (operation, step) in enumerate(zip(words[9::2], words[10::2])):
```

Comment:
I think it's not necessary, because there is not much happening in `_parse_cigar` and `_parse_vulgar`. so effectively those functions already play the role of a `calc_steps` function.